File: src/Sinbas.Domain/Lote.fs
namespace Sinbas.Domain

open System

type EstadoLote =
    | Activo
    | Agotado
    | Bloqueado
    | Archivado

type Lote =
    { Id: LoteId
      Codigo: CodigoLote
      ProductoId: ProductoId

      FechaIngreso: DateOnly

      UbicacionId: UbicacionId option

      Estado: EstadoLote

      Observaciones: string option }

module Lote =

    let estaActivo lote = lote.Estado = Activo

    let marcarAgotado lote = { lote with Estado = Agotado }

    let bloquear motivo lote =

        let observaciones =
            match lote.Observaciones with
            | None -> Some motivo

            | Some prev -> Some(sprintf "%s | %s" prev motivo)

        { lote with
            Estado = Bloqueado
            Observaciones = observaciones }

    let archivar lote = { lote with Estado = Archivado }

File: front/src/domain/models/Lote.ts
export interface LoteItem {
  id: string;
  codigo: string;
  productoId: string;
  nombreProducto: string;
  procedencia?: string;
  cantidadInicial: number;
  cantidadActual: number;
  unidad: string;
  fechaIngreso: string;
  ubicacion?: string;
  estado: 'Activo' | 'Agotado' | 'Bloqueado' | 'Archivado';
  observaciones?: string;
}

export interface CreateLotePayload {
  productoId: string;
  codigoPersonalizado?: string;
  procedencia?: string;
  cantidad: number;
  unidad: string;
  fechaIngreso: string;
  ubicacion?: string;
  observaciones?: string;
}

export interface BloquearLotePayload {
  motivo: string;
}
